// @graphprotocol/graph-cli/dist/commands/codegen.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const core_1 = require("@oclif/core");
const DataSourcesExtractor = __importStar(require("../command-helpers/data-sources"));
const version_1 = require("../command-helpers/version");
const debug_1 = __importDefault(require("../debug"));
const protocols_1 = __importDefault(require("../protocols"));
const type_generator_1 = __importDefault(require("../type-generator"));
const codegenDebug = (0, debug_1.default)('graph-cli:codegen');
class CodegenCommand extends core_1.Command {
    async run() {
        const { args: { 'subgraph-manifest': manifest }, flags: { 'output-dir': outputDir, 'skip-migrations': skipMigrations, watch, uncrashable, 'uncrashable-config': uncrashableConfig, }, } = await this.parse(CodegenCommand);
        codegenDebug('Initialized codegen manifest: %o', manifest);
        let protocol;
        try {
            // Checks to make sure codegen doesn't run against
            // older subgraphs (both apiVersion and graph-ts version).
            //
            // We don't want codegen to run without these conditions
            // because that would mean the CLI would generate code to
            // the wrong AssemblyScript version.
            await (0, version_1.assertManifestApiVersion)(manifest, '0.0.5');
            await (0, version_1.assertGraphTsVersion)(path_1.default.dirname(manifest), '0.25.0');
            const dataSourcesAndTemplates = await DataSourcesExtractor.fromFilePath(manifest);
            protocol = protocols_1.default.fromDataSources(dataSourcesAndTemplates);
        }
        catch (e) {
            this.error(e, { exit: 1 });
        }
        const generator = new type_generator_1.default({
            subgraphManifest: manifest,
            outputDir,
            skipMigrations,
            protocol,
            uncrashable,
            uncrashableConfig: uncrashableConfig || 'uncrashable-config.yaml',
        });
        // Watch working directory for file updates or additions, trigger
        // type generation (if watch argument specified)
        if (watch) {
            await generator.watchAndGenerateTypes();
        }
        else if (!(await generator.generateTypes())) {
            process.exitCode = 1;
        }
    }
}
exports.default = CodegenCommand;
CodegenCommand.description = 'Generates AssemblyScript types for a subgraph.';
CodegenCommand.args = {
    'subgraph-manifest': core_1.Args.string({
        default: 'subgraph.yaml',
    }),
};
CodegenCommand.flags = {
    help: core_1.Flags.help({
        char: 'h',
    }),
    'output-dir': core_1.Flags.directory({
        summary: 'Output directory for generated types.',
        char: 'o',
        default: 'generated/',
    }),
    'skip-migrations': core_1.Flags.boolean({
        summary: 'Skip subgraph migrations.',
    }),
    watch: core_1.Flags.boolean({
        summary: 'Regenerate types when subgraph files change.',
        char: 'w',
    }),
    uncrashable: core_1.Flags.boolean({
        summary: 'Generate Float Subgraph Uncrashable helper file.',
        char: 'u',
    }),
    'uncrashable-config': core_1.Flags.file({
        summary: 'Directory for uncrashable config.',
        aliases: ['uc'],
        // TODO: using a default sets the value and therefore requires --uncrashable
        // default: 'uncrashable-config.yaml',
        dependsOn: ['uncrashable'],
    }),
};
