// zksync-web3/build/src/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTypedDataSignatureCorrect = exports.isMessageSignatureCorrect = exports.getL2HashFromPriorityOp = exports.parseTransaction = exports.hashBytecode = exports.serialize = exports.checkBaseCost = exports.createAddress = exports.create2Address = exports.getDeployedContracts = exports.getHashedL2ToL1Msg = exports.layer1TxDefaults = exports.sleep = exports.isETH = exports.RECOMMENDED_GAS_LIMIT = exports.DEFAULT_ERGS_PER_PUBDATA_LIMIT = exports.EIP712_TX_TYPE = exports.EIP1271_MAGIC_VALUE = exports.L1_MESSENGER_ADDRESS = exports.CONTRACT_DEPLOYER_ADDRESS = exports.BOOTLOADER_FORMAL_ADDRESS = exports.L2_BRIDGE_ABI = exports.L1_BRIDGE_ABI = exports.IERC1271 = exports.IERC20 = exports.L1_MESSENGER = exports.CONTRACT_DEPLOYER = exports.ZKSYNC_MAIN_ABI = exports.ETH_ADDRESS = void 0;
const ethers_1 = require("ethers");
const types_1 = require("./types");
__exportStar(require("./paymaster-utils"), exports);
exports.ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ZKSYNC_MAIN_ABI = new ethers_1.utils.Interface(require('../../abi/IZkSync.json').abi);
exports.CONTRACT_DEPLOYER = new ethers_1.utils.Interface(require('../../abi/ContractDeployer.json').abi);
exports.L1_MESSENGER = new ethers_1.utils.Interface(require('../../abi/L1Messenger.json').abi);
exports.IERC20 = new ethers_1.utils.Interface(require('../../abi/IERC20.json').abi);
exports.IERC1271 = new ethers_1.utils.Interface(require('../../abi/IERC1271.json').abi);
exports.L1_BRIDGE_ABI = new ethers_1.utils.Interface(require('../../abi/IL1Bridge.json').abi);
exports.L2_BRIDGE_ABI = new ethers_1.utils.Interface(require('../../abi/IL2Bridge.json').abi);
exports.BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
exports.CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
exports.L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
exports.EIP1271_MAGIC_VALUE = '0x1626ba7e';
exports.EIP712_TX_TYPE = 0x71;
// Default pubdata price limit to be used if user provided none.
const GAS_PER_PUBDATA_BYTE = 16;
exports.DEFAULT_ERGS_PER_PUBDATA_LIMIT = GAS_PER_PUBDATA_BYTE * 10000;
exports.RECOMMENDED_GAS_LIMIT = {
    // TODO: make better estimates
    DEPOSIT: 600000,
    EXECUTE: 620000,
    ERC20_APPROVE: 50000
};
function isETH(token) {
    return token.toLowerCase() == exports.ETH_ADDRESS;
}
exports.isETH = isETH;
function sleep(millis) {
    return new Promise((resolve) => setTimeout(resolve, millis));
}
exports.sleep = sleep;
function layer1TxDefaults() {
    return {
        queueType: types_1.PriorityQueueType.Deque,
        opTree: types_1.PriorityOpTree.Full
    };
}
exports.layer1TxDefaults = layer1TxDefaults;
function getHashedL2ToL1Msg(sender, msg) {
    const encodedMsg = new Uint8Array([
        ...ethers_1.ethers.utils.arrayify(exports.L1_MESSENGER_ADDRESS),
        ...ethers_1.ethers.utils.zeroPad(sender, 32),
        ...ethers_1.ethers.utils.arrayify(ethers_1.ethers.utils.keccak256(msg))
    ]);
    return ethers_1.ethers.utils.sha256(encodedMsg);
}
exports.getHashedL2ToL1Msg = getHashedL2ToL1Msg;
function getDeployedContracts(receipt) {
    const addressBytesLen = 40;
    const deployedContracts = receipt.logs
        .filter((log) => log.topics[0] == ethers_1.utils.id('ContractDeployed(address,bytes32,address)') &&
        log.address == exports.CONTRACT_DEPLOYER_ADDRESS)
        // Take the last topic (deployed contract address as U256) and extract address from it (U160).
        .map((log) => {
        const sender = `0x${log.topics[1].slice(log.topics[1].length - addressBytesLen)}`;
        const bytesCodehash = log.topics[2];
        const address = `0x${log.topics[3].slice(log.topics[3].length - addressBytesLen)}`;
        return {
            sender: ethers_1.utils.getAddress(sender),
            bytecodeHash: bytesCodehash,
            deployedAddress: ethers_1.utils.getAddress(address)
        };
    });
    return deployedContracts;
}
exports.getDeployedContracts = getDeployedContracts;
function create2Address(sender, bytecodeHash, salt, input) {
    const prefix = ethers_1.ethers.utils.keccak256(ethers_1.ethers.utils.toUtf8Bytes('zksyncCreate2'));
    const inputHash = ethers_1.ethers.utils.keccak256(input);
    const addressBytes = ethers_1.ethers.utils
        .keccak256(ethers_1.ethers.utils.concat([prefix, ethers_1.ethers.utils.zeroPad(sender, 32), salt, bytecodeHash, inputHash]))
        .slice(26);
    return ethers_1.ethers.utils.getAddress(addressBytes);
}
exports.create2Address = create2Address;
function createAddress(sender, senderNonce) {
    const prefix = ethers_1.ethers.utils.keccak256(ethers_1.ethers.utils.toUtf8Bytes('zksyncCreate'));
    const addressBytes = ethers_1.ethers.utils
        .keccak256(ethers_1.ethers.utils.concat([
        prefix,
        ethers_1.ethers.utils.zeroPad(sender, 32),
        ethers_1.ethers.utils.zeroPad(ethers_1.ethers.utils.hexlify(senderNonce), 32)
    ]))
        .slice(26);
    return ethers_1.ethers.utils.getAddress(addressBytes);
}
exports.createAddress = createAddress;
async function checkBaseCost(baseCost, value) {
    if (baseCost.gt(await value)) {
        throw new Error(`The base cost of performing the priority operation is higher than the provided value parameter ` +
            `for the transaction: baseCost: ${baseCost}, provided value: ${value}`);
    }
}
exports.checkBaseCost = checkBaseCost;
function serialize(transaction, signature) {
    var _a;
    if (transaction.customData == null && transaction.type != exports.EIP712_TX_TYPE) {
        return ethers_1.utils.serializeTransaction(transaction, signature);
    }
    if (!transaction.chainId) {
        throw Error("Transaction chainId isn't set");
    }
    function formatNumber(value, name) {
        const result = ethers_1.utils.stripZeros(ethers_1.BigNumber.from(value).toHexString());
        if (result.length > 32) {
            throw new Error('invalid length for ' + name);
        }
        return result;
    }
    if (!transaction.from) {
        throw new Error('Explicitly providing `from` field is reqiured for EIP712 transactions');
    }
    const from = transaction.from;
    const meta = transaction.customData;
    let maxFeePerErg = transaction.maxFeePerGas || transaction.gasPrice || 0;
    let maxPriorityFeePerErg = transaction.maxPriorityFeePerGas || maxFeePerErg;
    const fields = [
        formatNumber(transaction.nonce || 0, 'nonce'),
        formatNumber(maxPriorityFeePerErg, 'maxPriorityFeePerGas'),
        formatNumber(maxFeePerErg, 'maxFeePerGas'),
        formatNumber(transaction.gasLimit || 0, 'gasLimit'),
        transaction.to != null ? ethers_1.utils.getAddress(transaction.to) : '0x',
        formatNumber(transaction.value || 0, 'value'),
        transaction.data || '0x'
    ];
    if (signature) {
        const sig = ethers_1.utils.splitSignature(signature);
        fields.push(formatNumber(sig.recoveryParam, 'recoveryParam'));
        fields.push(ethers_1.utils.stripZeros(sig.r));
        fields.push(ethers_1.utils.stripZeros(sig.s));
    }
    else {
        fields.push(formatNumber(transaction.chainId, 'chainId'));
        fields.push('0x');
        fields.push('0x');
    }
    fields.push(formatNumber(transaction.chainId, 'chainId'));
    fields.push(ethers_1.utils.getAddress(from));
    // Add meta
    fields.push(formatNumber(meta.ergsPerPubdata || exports.DEFAULT_ERGS_PER_PUBDATA_LIMIT, 'ergsPerPubdata'));
    fields.push(((_a = meta.factoryDeps) !== null && _a !== void 0 ? _a : []).map((dep) => ethers_1.utils.hexlify(dep)));
    if (meta.customSignature && ethers_1.ethers.utils.arrayify(meta.customSignature).length == 0) {
        throw new Error('Empty signatures are not supported');
    }
    fields.push(meta.customSignature || '0x');
    if (meta.paymasterParams) {
        fields.push([meta.paymasterParams.paymaster, ethers_1.ethers.utils.hexlify(meta.paymasterParams.paymasterInput)]);
    }
    else {
        fields.push([]);
    }
    return ethers_1.utils.hexConcat([[exports.EIP712_TX_TYPE], ethers_1.utils.RLP.encode(fields)]);
}
exports.serialize = serialize;
function hashBytecode(bytecode) {
    // For getting the consistent length we first convert the bytecode to UInt8Array
    const bytecodeAsArray = ethers_1.ethers.utils.arrayify(bytecode);
    const hashStr = ethers_1.ethers.utils.sha256(bytecodeAsArray);
    const hash = ethers_1.ethers.utils.arrayify(hashStr);
    // Note that the length of the bytecode
    // should be provided in 32-byte words.
    const bytecodeLength = ethers_1.ethers.utils.arrayify(bytecodeAsArray.length / 32);
    if (bytecodeLength.length > 2) {
        throw new Error('Bytecode length must be less than 2^16 bytes');
    }
    // The bytecode should always take the first 2 bytes of the bytecode hash,
    // so we pad it from the left in case the length is smaller than 2 bytes.
    const bytecodeLengthPadded = ethers_1.ethers.utils.zeroPad(bytecodeLength, 2);
    hash.set(bytecodeLengthPadded, 0);
    return hash;
}
exports.hashBytecode = hashBytecode;
function parseTransaction(payload) {
    function handleAddress(value) {
        if (value === '0x') {
            return null;
        }
        return ethers_1.utils.getAddress(value);
    }
    function handleNumber(value) {
        if (value === '0x') {
            return ethers_1.BigNumber.from(0);
        }
        return ethers_1.BigNumber.from(value);
    }
    function arrayToPaymasterParams(arr) {
        if (arr.length == 0) {
            return undefined;
        }
        if (arr.length != 2) {
            throw new Error(`Invalid paymaster parameters, expected to have length of 2, found ${arr.length}`);
        }
        return {
            paymaster: ethers_1.utils.getAddress(arr[0]),
            paymasterInput: ethers_1.utils.arrayify(arr[1])
        };
    }
    const bytes = ethers_1.utils.arrayify(payload);
    if (bytes[0] != exports.EIP712_TX_TYPE) {
        return ethers_1.utils.parseTransaction(bytes);
    }
    const raw = ethers_1.utils.RLP.decode(bytes.slice(1));
    const transaction = {
        type: exports.EIP712_TX_TYPE,
        nonce: handleNumber(raw[0]).toNumber(),
        maxPriorityFeePerGas: handleNumber(raw[1]),
        maxFeePerGas: handleNumber(raw[2]),
        gasLimit: handleNumber(raw[3]),
        to: handleAddress(raw[4]),
        value: handleNumber(raw[5]),
        data: raw[6],
        chainId: handleNumber(raw[10]),
        from: handleAddress(raw[11]),
        customData: {
            ergsPerPubdata: handleNumber(raw[12]),
            factoryDeps: raw[13],
            customSignature: raw[14],
            paymasterParams: arrayToPaymasterParams(raw[15])
        }
    };
    const ethSignature = {
        v: handleNumber(raw[7]).toNumber(),
        r: raw[8],
        s: raw[9]
    };
    if ((ethers_1.utils.hexlify(ethSignature.r) == '0x' || ethers_1.utils.hexlify(ethSignature.s) == '0x') &&
        !transaction.customData.customSignature) {
        return transaction;
    }
    transaction.hash = ethers_1.utils.keccak256(payload);
    if (ethSignature.v !== 0 && ethSignature.v !== 1 && !transaction.customData.customSignature) {
        throw new Error('Failed to parse signature');
    }
    return transaction;
}
exports.parseTransaction = parseTransaction;
function getL2HashFromPriorityOp(txReceipt, zkSyncAddress) {
    let txHash = null;
    for (const log of txReceipt.logs) {
        if (log.address.toLowerCase() != zkSyncAddress.toLowerCase()) {
            continue;
        }
        try {
            const priorityQueueLog = exports.ZKSYNC_MAIN_ABI.parseLog(log);
            if (priorityQueueLog && priorityQueueLog.args.txHash != null) {
                txHash = priorityQueueLog.args.txHash;
            }
        }
        catch { }
    }
    if (!txHash) {
        throw new Error('Failed to parse tx logs');
    }
    return txHash;
}
exports.getL2HashFromPriorityOp = getL2HashFromPriorityOp;
// The method with similar functionality is already available in ethers.js,
// the only difference is that we provide additional `try { } catch { }`
// for error-resilience.
//
// It will also pave the road for allowing future EIP-1271 signature verification, by
// letting our SDK have functionality to verify signatures.
function isECDSASignatureCorrect(address, msgHash, signature) {
    try {
        return address == ethers_1.ethers.utils.recoverAddress(msgHash, signature);
    }
    catch {
        // In case ECDSA signature verification has thrown an error,
        // we simply consider the signature as incorrect.
        return false;
    }
}
async function isEIP1271SignatureCorrect(provider, address, msgHash, signature) {
    const accountContract = new ethers_1.ethers.Contract(address, exports.IERC1271, provider);
    // This line may throw an exception if the contract does not implement the EIP1271 correctly.
    // But it may also throw an exception in case the internet connection is lost.
    // It is the caller's responsibility to handle the exception.
    const result = await accountContract.isValidSignature(msgHash, signature);
    return result == exports.EIP1271_MAGIC_VALUE;
}
async function isSignatureCorrect(provider, address, msgHash, signature) {
    let isContractAccount = false;
    const code = await provider.getCode(address);
    isContractAccount = ethers_1.ethers.utils.arrayify(code).length != 0;
    if (!isContractAccount) {
        return isECDSASignatureCorrect(address, msgHash, signature);
    }
    else {
        return await isEIP1271SignatureCorrect(provider, address, msgHash, signature);
    }
}
// Returns `true` or `false` depending on whether or not the account abstraction's
// signature is correct. Note, that while currently it does not do any `async` actions.
// in the future it will. That's why the `Promise<boolean>` is returned.
async function isMessageSignatureCorrect(provider, address, message, signature) {
    const msgHash = ethers_1.ethers.utils.hashMessage(message);
    return await isSignatureCorrect(provider, address, msgHash, signature);
}
exports.isMessageSignatureCorrect = isMessageSignatureCorrect;
// Returns `true` or `false` depending on whether or not the account abstraction's
// EIP712 signature is correct. Note, that while currently it does not do any `async` actions.
// in the future it will. That's why the `Promise<boolean>` is returned.
async function isTypedDataSignatureCorrect(provider, address, domain, types, value, signature) {
    const msgHash = ethers_1.ethers.utils._TypedDataEncoder.hash(domain, types, value);
    return await isSignatureCorrect(provider, address, msgHash, signature);
}
exports.isTypedDataSignatureCorrect = isTypedDataSignatureCorrect;
